core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/dimension/legacy/ComplexDimensionIndexCodec.java

```diff
@@ -46,12 +47,13 @@ public ColumnPageEncoder createEncoder(Map<String, String> parameter) {
     return new IndexStorageEncoder() {
       @Override
       void encodeIndexStorage(ColumnPage inputPage) {
-        BlockIndexerStorage<byte[][]> indexStorage =
-            new BlockIndexerStorageForShort(inputPage.getByteArrayPage(), false, false, false);
-        byte[] flattened = ByteUtil.flatten(indexStorage.getDataPage());
+        BlockIndexerStorage<ByteBuffer[]> indexStorage =
+            new BlockIndexerStorageForShort(inputPage.getByteBufferArrayPage(false),
+                false, false, false);
+        ByteBuffer flattened = ByteUtil.flatten(indexStorage.getDataPage());
         Compressor compressor = CompressorFactory.getInstance().getCompressor(
             inputPage.getColumnCompressorName());
-        byte[] compressed = compressor.compressByte(flattened);
+        byte[] compressed = ByteUtil.byteBufferToBytes(compressor.compressByte(flattened));
```

Comment:
I think we have converted 2D byte array to bytebuffer, to support compression directly on the byte buffer.
But now we have to convert to byte[] again !, possible to keep bytebuffer itself ?

so compression gc might have reduced, but decompression gc might have increased now. Please compare the before and after performance and memory usage.